### PR TITLE
Remove Commons HttpClient 3.x

### PR DESCRIPTION
**Pull request by basil**
### Problem

Core still bundles a [patched version](https://github.com/jenkinsci/lib-commons-httpclient) of the deprecated [Commons HttpClient 3.x](https://hc.apache.org/httpclient-legacy/) library for use by plugins. This frequently confuses security scanners and is a maintenance liability.

### Solution

Create a new Commons HttpClient 3.x API plugin and remove this library from Jenkins core.

### Implementation

While we considered using the detached plugin mechanism, there are not too many consumers, and most of them are very old. Furthermore many of them are already bundling HttpClient 3.x. It seemed simpler to update the remainder to include an explicit dependency on the HttpClient 3.x API plugin, which can be done before this PR is released to ensure a smooth transition. Therefore I plan to keep this PR on hold until a critical threshold of plugins are depending on the HttpClient 3.x API plugin. I plan to create a Jira epic to track the slow progress on this long tail of plugins. When we feel that a sufficient number of popular plugins have been released, we can merge and release this PR.

The solution for each plugin can be chosen on a case-by-case basis. In the ideal case the plugin can be updated to use the native Java 11 HTTP client and eliminate any dependency on a third-party library. In some cases, the plugin was just using the `org.apache.commons.httpclient.HttpStatus` and can be trivially migrated to `java.net.HttpURLConnection` as in https://github.com/jenkinsci/docker-build-step-plugin/pull/84. In the worst case, adding an explicit dependency on the forthcoming HttpClient 3.x API is the simplest way to keep the plugin working as it already does without regression.

### Testing done

To test this change, I built a custom plugin that consumes HttpClient 3.x to fetch https://www.google.com and verified that it worked correctly against 2.375 and failed with this PR as expected. Then I added a dependency on the new HttpClient 3.x API plugin and verified that the custom plugin started working again with this PR. I also verified that the custom plugin with a dependency on the new HttpClient 3.x API plugin worked correctly on Jenkins 2.375 (without this PR), demonstrating that it is safe to release the HttpClient 3.x API plugin and start adding dependencies to it even before HttpClient 3.x is removed from core in this PR.

### Proposed changelog entries

Jenkins no longer bundles a [patched version](https://github.com/jenkinsci/lib-commons-httpclient) of the deprecated [Commons HttpClient 3.x](https://hc.apache.org/httpclient-legacy/) library for use by plugins. Plugins should be migrated to the native Java 11 HTTP client or updated to depend on the legacy Commons HttpClient 3.x API plugin.

### Proposed upgrade guidelines

Jenkins no longer bundles a [patched version](https://github.com/jenkinsci/lib-commons-httpclient) of the deprecated [Commons HttpClient 3.x](https://hc.apache.org/httpclient-legacy/) library for use by plugins. Plugins should be migrated to the native Java 11 HTTP client or updated to depend on the legacy Commons HttpClient 3.x API plugin.

### Submitter checklist

- [ ] The Jira issue, if it exists, is well-described.
- [ ] The changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developers, depending on the change) and are in the imperative mood (see [examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)).
  - Fill in the **Proposed upgrade guidelines** section only if there are breaking changes or changes that may require extra steps from users during upgrade.
- [ ] There is automated testing or an explanation as to why this change has no tests.
- [ ] New public classes, fields, and methods are annotated with `@Restricted` or have `@since TODO` Javadocs, as appropriate.
- [ ] New deprecations are annotated with `@Deprecated(since = "TODO")` or `@Deprecated(forRemoval = true, since = "TODO")`, if applicable.
- [ ] New or substantially changed JavaScript is not defined inline and does not call `eval` to ease future introduction of Content Security Policy (CSP) directives (see [documentation](https://www.jenkins.io/doc/developer/security/csp/)).
- [ ] For dependency updates, there are links to external changelogs and, if possible, full differentials.
- [ ] For new APIs and extension points, there is a link to at least one consumer.

### Desired reviewers

@mention

### Maintainer checklist

Before the changes are marked as `ready-for-merge`:

- [ ] There are at least two (2) approvals for the pull request and no outstanding requests for change.
- [ ] Conversations in the pull request are over, or it is explicit that a reviewer is not blocking the change.
- [ ] Changelog entries in the pull request title and/or **Proposed changelog entries** are accurate, human-readable, and in the imperative mood.
- [ ] Proper changelog labels are set so that the changelog can be generated automatically.
- [ ] If the change needs additional upgrade steps from users, the `upgrade-guide-needed` label is set and there is a **Proposed upgrade guidelines** section in the pull request title (see [example](https://github.com/jenkinsci/jenkins/pull/4387)).
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins.io/issues/?filter=12146)).

<a href="https://gitpod.io/#https://github.com/jenkinsci/jenkins/pull/7312"></a>

### CODE REVIEW

**Comment by basil** [comment]
The failing tests are due to the bundled `jdk-tool` plugin depending on HttpClient 3.x. That dependency will need to be eliminated before this work can continue.

File: war/pom.xml
Comment (basil) [inline war/pom.xml:462]: Failing tests are fixed with the inclusion of https://github.com/jenkinsci/jdk-tool-plugin/pull/49

**Review by NotMyFault — APPROVED** [review]

**Comment by basil** [comment]
I am ready to proceed with shipping this in a weekly release. [This spreadsheet](https://docs.google.com/spreadsheets/d/1jXTJmHLVg6HpWA_-8vkcmL0Q005EKqIkykPgRLsj9K0/edit) tracks progress in adopting plugins; any plugin with significant usage has already been prepared and released. I plan to see if any more fixes trickle in during the time between now and the next LTS release, possibly fixing a few more long-tail plugins if needed, and then formally deprecating the rest in the Update Center around the time we ship the next LTS release.

**Review by oleg-nenashev — APPROVED** [review]
As someone who created this library, I strongly support removing it. Removing it in a few months was the plan all way along when I was implementing it, but the effort was stopped in the middle due to the change of the team's priorities and the reorg. 

If we were in 2018(?), I would have suggested making the Commons HttpClient 3.x an API plugin. But there is no point in that in 2022, especially since the interfaces in https://github.com/jenkinsci/lib-commons-httpclient have been deprecated since many years ago.

My concern, before doing a merge, is removal of two classes from the Jenkins core. These classes are a part of API, and they have not been deprecated. They clearly extend HTTPClient 3.x and hence their binary compatibility cannot be maintained, but I think we can live with that since the new API plugin provides it

Thanks @basil !

**Comment by basil** [comment]
This PR is now ready for merge. We will merge it after approximately 24 hours if there is no negative feedback. Please see the [merge process documentation](https://github.com/jenkinsci/jenkins/blob/master/docs/MAINTAINERS.adoc#merge-process) for more information about the merge process. Thanks!

**Review by MarkEWaite — APPROVED** [review]

**Pull request merged**
